functions.go

```diff
@@ -328,3 +328,49 @@ func errorStack(err error) []string {
 	}
 	return result
 }
+
+// CauseEquals tests, if this or any cause of this error is equal to another error, or it's cause
+//
+// For example:
+//		ErrNotFound := errors.New("Not found")
+//		inSome := errors.Trace(ErrNotFound)
+//		veryDeep := errors.Trace(inSome)
+//		returned := errors.Trace(veryDeep)
+//		if errors.CauseEquals(ErrNotFound, returned) {
+//			fmt.Println("Can return soft error 404")
+//		} else {
+//			fmt.Println("Serious error, returning with code 500")
+//		}
+func CauseEquals(err1, err2 error) bool {
```

Comment:
errors.Cause(err) is constructed so that it works with normal errors and errors from the juju/errors package.

https://github.com/juju/errors/blob/master/functions.go#L188

I tweaked your example to show what I mean:
```go
package main

import (
	"fmt"

	"github.com/juju/errors"
)

var ErrNotFound = errors.New("Resource not found")

func readElement() error {
	return ErrNotFound
}

func complexReadElement() error {
	return errors.Trace(ErrNotFound)
}

func refactoredReadElement() error {
	if err := middlewareStuff(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func middlewareStuff() error {
	if err := lowLevelDatabaseRead(); err != nil {
		return errors.Trace(err)
	}
	return nil
}

func lowLevelDatabaseRead() error {
	return ErrNotFound
}

func annotatedRead() error {
	if err := middlewareStuff(); err != nil {
		return errors.Annotate(err, "Something")
	}
	return nil
}

func annotatedRead2() error {
	if err := annotatedRead(); err != nil {
		return errors.Annotate(err, "Something 2")
	}
	return nil
}

func main() {
	if errors.Cause(readElement()) == ErrNotFound {
		fmt.Println("Soft error 1")
	} else {
		fmt.Println("EXPECTING: Soft error 1")
	}

	if errors.Cause(complexReadElement()) == ErrNotFound {
		fmt.Println("Soft error 2")
	} else {
		fmt.Println("EXPECTING: Soft error 2")
	}

	// I do not want casting, but it is ok
	if errors.Cause(complexReadElement()) == ErrNotFound {
		fmt.Println("Soft error 3")
	} else {
		fmt.Println("EXPECTING: Soft error 3")
	}

	// Ok. Did not expect this :)
	if errors.Cause(refactoredReadElement()) == ErrNotFound {
		fmt.Println("Soft error 4")
	} else {
		fmt.Println("EXPECTING: Soft error 4")
	}

	// Ok. We could cast it and get what I expected.
	if errors.Cause(refactoredReadElement()) == errors.Cause(middlewareStuff()) {
		fmt.Println("Soft error 5")
	} else {
		fmt.Println("EXPECTING: Soft error 5")
	}

	// Ok. We could cast it and get what I expected.
	if errors.Cause(annotatedRead2()) == errors.Cause(annotatedRead()) {
		fmt.Println("Soft error 6")
	} else {
		fmt.Println("EXPECTING: Soft error 6")
	}
}
```
```
Soft error 1
Soft error 2
Soft error 3
Soft error 4
Soft error 5
Soft error 6
```